aces/Loris-MRI
# Add support for multiple MRI protocols.

### Description

This PR adds the possibility to have multiple MRI protocols for a given study and multiple groups of MRI protocol checks. You can now group scan type specifications listed in `mri_protocol` and have specific groups taken into account __or__ ignored based on the scan properties (visit at which the scan was taken, etc...). Similarly, you can now group lines of the `mri_protocol_check` and specify under which conditions specific groups are taken into account __or__ ignored for a given scan. In a nutshell, this grouping feature specifies which lines of table `mri_protocol` and `mri_protocol_checks` are active or inactive for a given scan. 

### How to define new MRI protocol groups and MRI protocol checks groups

First you have to run the SQL patch that updates the database schema for this new feature. Once you have done this, the behaviour of the pipeline should not have changed: all the lines of the `mri_protocol` and `mri_protocol_checks` will be active for all scans . To create a new MRI protocol group:

**Step 1:** Insert a record in `mri_protocol_group` with the new group name and ID.  

**Step 2:** Associate each line in table `mri_protocol` to a protocol group in `mri_protocol_group` by filling out column`MriProtocolGroupID`.

**Step 3:** Determine the conditions under which the new group will be used to identify the type of a scan. Conditions include the project ID, sub-project ID or visit label of the session at which the scan was done. There are columns in table `mri_protocol_group_target` for each of these types of conditions. When more than one column is filled in this table, the conditions are **AND**ed together.

The steps to create a new MRI protocol checks group are analog to those described above but use tables `mri_protocol_checks_group` and `mri_protocol_checks_group_target` instead.

Each line of the `mri_protocol` table belongs to either one group or no group at all, and this will be enforced in the MySQL schema. Lines of `mri_protocol_checks` can belong to 0, 1 or many protocol checks group though.

In addition to creating the new tables described above, this PR also modifies table `mri_violated_scans` and adds a new column that contains the MRI protocol group that was used when trying to identify the scan type. When the MRI cannot determine the type of a scan because the scan does not match any of the conditions listed in `mri_protocol_group_target`, the value in this new column will be set to `NULL`. 

### How to test

Make sure you have sourced the SQL patch that extends the database in order to support multiple MRI protocols. Create new MRI protocol groups and new groups of MRI protocol check and make sure that the correct groups are used by the MRI pipeline according to the scan properties.

### See also

Associated Redmine ticket: https://redmine.cbrain.mcgill.ca/issues/15452
Associated PR on the LORIS side: [https://github.com/aces/Loris/pull/4321](https://github.com/aces/Loris/pull/4321)
